File: 101-200/leetcode#118.cpp
class Solution {
public:
    vector<vector<int>> generate(int numRows) {
        int n = numRows;
        vector<vector<int>> res;
        
        if (n == 1) {
            res.push_back({1});
        } else if (n == 2) {
            res.push_back({1});
            res.push_back({1, 1});
        } else {
            res.push_back({1});
            res.push_back({1, 1});
            int level = 2;
            for (int i = 1; i < n - 1; i++) {
                vector<int> v;
                for (int j = 0; j <= level; j++) {
                    if (j == 0)
                        v.push_back(res[i][0]);
                    else if (j == level)
                        v.push_back(res[i][level-1]);
                    else
                        v.push_back(res[i][j-1] + res[i][j]);
                }
                level++;
                res.push_back(v);
            }
        }
        
        return res;
    }
};

File: 101-200/leetcode#118.java
class Solution {
    public List<List<Integer>> generate(int numRows) {
        List<List<Integer>> res = new ArrayList<>();

        //no element -> return empty list
        if (numRows == 0) return res;

        //for only one element
        List<Integer> firstRow = new ArrayList<>();
        firstRow.add(1);
        res.add(firstRow);
        if (numRows == 1) return res;

        //for n number of rows
        for (int i = 1; i < numRows; i++) {
            List<Integer> prevList = res.get(i-1);

            List<Integer> row = new ArrayList<>();
            
            //start element
            row.add(1);

            for (int j = 0; j < i-1; j++) {
                row.add(prevList.get(j) + prevList.get(j+1));
            }

            //end element
            row.add(1);

            res.add(row);
        }

        return res;
    }
}

//time comlexity - O(N^2)
//space complexity - O(N^2)
